segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/SyncStorage.java

```diff
@@ -216,6 +216,40 @@ default SegmentHandle create(String streamSegmentName, SegmentRollingPolicy roll
      */
     boolean supportsTruncation();
 
+    /**
+     * Gets a value indicating whether this {@link SyncStorage} implementation can replace whole Segments with new contents.
+     *
+     * @return True or false.
+     */
+    default boolean supportsReplace() {
+        return false;
+    }
+
+    /**
+     * Replaces a Segment with the given contents.  Please refer to the actual implementing class for more details with
+     * respect to behavior, atomicity and recovery mechanisms.
+     *
+     * @param segment A {@link SegmentHandle} representing the Segment to replace.
+     * @param contents A {@link BufferView} representing the new contents of the Segment.
+     * @throws StreamSegmentNotExistsException When the given Segment does not exist in Storage.
+     * @throws StorageNotPrimaryException      When this Storage instance is no longer primary for this Segment (it was
+     *                                         fenced out).
+     * @throws UnsupportedOperationException   If {@link #supportsReplace()} returns false.
+     */
+    default void replace(SegmentHandle segment, BufferView contents) throws StreamSegmentException {
+        throw new UnsupportedOperationException("replace() is not implemented");
+    }
+
+    /**
+     * Returns a new {@link SyncStorage} instance for the same Storage type as this one, but with {@link #replace} support
+     * enabled. If there is no such implementation, this instance is returned.
+     *
+     * @return Either this instance or a new {@link SyncStorage} instance based on this one that can perform replaces.
+     */
+    default SyncStorage withReplaceSupport() {
```

Comment:
This is the only call that feels odd when comparing to the other calls in the interface. To be able to get rid of it, we would need to make `FileSystemStorage` support it directly rather than have a derived class `FileSystemStorageWithReplace`, which seems possible and it is not clear to me whether it is incorrect or simply undesirable.

Reply:
`FileSystemWithReplace` does a bit of extra work for every file access and catches certain exceptions. To prevent that overhead from happening **on every single LTS call** which could negatively affect performance (and open the door for more hacks in the future (since it's easy to use such functionality once there), I decided to explicitly offer this as a separate piece of functionality. The only place where this functionality is required is in `RollingStorage` for the header file. Furthermore, `SyncStorage` is the internal LTS contract, not the external one (that one is `Storage`) so nobody else outside of `RollingStorage` would even have access to this. 

This is not a long-term solution. We are doing it as a short-term fix until SLTS will replace `RollingStorage` (SLTS does not have the problem that this PR is fixing). 

In other words, I'm deliberately trying to make it hard for people to accidentally use this feature and not slip through future code reviews unnoticed. We do not plan to support this functionality long term.